pkg/policy/mapstate.go

```diff
@@ -84,24 +85,124 @@ type MapState interface {
 
 // mapState is a state of a policy map.
 type mapState struct {
-	allows mapStateMap
-	denies mapStateMap
+	allows *mapStateMap
+	denies *mapStateMap
+}
+
+// Identities is a convenience interface for looking up CIDRs
+// associated with an identity
+type Identities interface {
+	GetNetsLocked(identity.NumericIdentity) []*net.IPNet
 }
 
 // mapStateMap is a convience type representing the actual structure mapping
 // policymap keys to policymap entries.
-type mapStateMap map[Key]MapStateEntry
+//
+// The `bitlpm.Trie` indexes the TrafficDirection, Protocol, and Port of
+// a policy Key but does **not** index the identity. Instead identities
+// that share TrafficDirection, Protocol, and Port are indexed in a builtin
+// map type that is the associated value of the key-prefix of TrafficDirection,
+// Protocol, and Port. This is done so that Identity does not explode
+// the size of the Trie. Consider the case of a policy that selects
+// many identities. In this case, if Identity was indexed then every
+// identity associated with the policy would create at least one
+// intermediate node in the Trie with its own sub node associated with
+// TrafficDirection, Protocol, and Port. When identity is not indexed
+// then one policy will map to one key-prefix with a builtin map type
+// that associates each identity with a MapStateEntry. This strategy
+// greatly enhances the usefuleness of the Trie and improves lookup,
+// deletion, and insertion times.
+type mapStateMap struct {
+	// t is a Trie that indexes policy Keys without their indentity
+	// and stores the idenitites in an associated builtin map with
```

Comment:
nit: If identity is disregarded, then maybe there should be a subtype that excludes Identity, composed into `Key`. The `Trie` would use this new key subtype. That way, the lookup in this specific Trie type would be clearer (no need to explain that identity is excluded if the identity isn't part of the key type used for indexing) and we could skip the corresponding zeroing logic and memory usage. The lookup functions into `mapStateMap` would continue to use the full `Key`.

Additionally:

```suggestion
	// t is a Trie that indexes policy Keys without their identity
	// and stores the identities in an associated builtin map with
```